### ISSUE TITLE
delayed quit for followers and leachers

### ISSUE THREAD

**Issue by alfton** [kolton/d2bot-with-kolbot#2002]
I know this issue have been up, and there are current some solutions typed in, but non of them work for me, the follower, and leacher just quick the game fast as the leader. And always get the bot in trouble with failed to join game.

Have somone have a good solutions they wanna share with me?

**Comment by AnotherCasualty** [comment]
I'm testing a solution that I've found in a previous thread. If it works I'll update here.

**Comment by alfton** [comment]
yeah, I have by the way tryed this solutions:
https://github.com/kolton/d2bot-with-kolbot/issues/1462#issuecomment-484983690

and the tips about puting delay near the end of baalhelper script
https://github.com/kolton/d2bot-with-kolbot/issues/1837#issuecomment-520651891

I have tryed all of this, no luck on it..

**Comment by AbbyRoad** [comment]
Method is working for me.  I still get a FTJ every couple of hours, but the change I made to the Bo script prevents (so far) a RD from occuring.

See Issue #1997

**Comment by alfton** [comment]
ahhh, ty, will try implementing that code when im home.. reporting back..

**Comment by alfton** [comment]
ill tryed to implement https://github.com/kolton/d2bot-with-kolbot/issues/1997#issuecomment-557913739
but no luck there. Follower leaving exactly same time as leader

**Comment by AbbyRoad** [comment]
After running just under 24 hrs, I just got hit with a 3 day RD.  During the time I was running, the average game length was 7.5 minutes.  I had 12 FTJ events in 24 hrs, each lasting about 15 minutes.  With the change to Bo bot code, the "waiting" game lengths were close to 7.5 minutes (ie. no different than a normal game).  The bot that FTJ'd did unsuccessfully attempt to enter the game 10 - 15 times during each of these periods.  I have no idea if this counts against the allowable games/hr count.

Prior to the recent uptick in FTJs, I was running 24/7 for about a month with 4 bots in private games that, on average, were about 6 minutes long.  Extremely rare FTJs and no extended period RDs.

In my current games, 3 bots are doing - AT, N, D and B, while the 4th does A1 - A3 chests.  The fourth bot is usually waiting in town for 10 - 30 secs while the other 3 are completing their run.  The bot idling in town does leave the game significantly later than the other 3.

Only the two follower bots have had FTJs (1 had 3, and the other 9).  I assume that this is due to delayed leaving code not working as well (or at all) for them.  (ie. if the bots are in town, the delay does work, but not if the bot is not sitting in town (in the Throne Room, for example).

So, to get the delay code to work properly, I believe it is necessary for the bots to be in town at the end of the run.  I'm not certain how to do get the bots to return to town before leaving.

Secondly, I don't know if Blizzard is cracking down on the number of games allowed per hour from a single internet connection, or if my recent use of a delay in the entry of the followers to the game is causing Blizzard to change how it does the count (ie. 3 followers joining more or less at the same time count as "1 additional game", while 3 followers joining in two second increments count as "3 additional games" towards the hourly allowable.)  As a minimum, I'll be removing the game entry stagger as it does not seem to affect the FTJ situation.

I hope this helps someone with way more experience than I to develop a solution.

**Comment by alfton** [comment]
well at the moment I have had 4 bot runing with out any problem. No realm down, or fail to join game. for 3 day in a row. Well the game are for over 20 min.  
And when the bot crashing, I have set a very long delay befor it rejoin: rand(900, 1200), and create new game is on rand(120, 160). The only thing now is to solve this leaving game at the same time if I have 3 of the bot in same game. So hoping someone find a good solutions to this problem.
then I can do some leveling up for one of my chars in my own runs.

**Comment by AbbyRoad** [comment]
4 bots x 3 runs/hr = 12 runs which is << 20 max people say is the limit.

Have you tried to add "Town.doChores();" as the last step in your run?  If you can get the bots back in town before exiting, then - based on my experience - the delay for leaving should work.

I'd suggest a the following Test process.  

First,  have two bots in a game, but leave one idling in town.  Confirm my experience that the idling bot leaves after the delay you coded in, when the leader leaves.

If that worked, and if you are doing Baal as the final step, then I'd suggest adding "Town.doChores();" just above the "return true;" at the end of the Baal.js file.

That should send the leader back to town before it exits.

If that worked, then allow the follower bot to tag along with the leader.  They should both return to town at the end of the run, before exiting.

Let me know if it works, or if there is a better place/file to put the "Town.doChores();" line of code.

**Comment by alfton** [comment]
well, the leader get in town, and sell the iten he picket up, and then he leave,. but follower, and leacher leaving game at the same time, tryed to put the Town.doChores in baal, autobaal, and the baalassistant, but no luck there, but we closing inn on something here..

**Comment by AbbyRoad** [comment]
Hmmm.  In my runs the follower that had the delayed leave time was sitting in A4 town, while the other 3 (leader & 2 followers) were in A5.  In your runs are all 3 chars in the same Act town?  I wouldn't think that would be an issue, but who knows.  When my RD ends, I'll do some testing.  Hopefully, someone with way more programing knowledge than I will show us the way!  Must be something obvious we are missing.

Good Luck!

**Comment by alfton** [comment]
yeah, Ill put my bot on halt at the moment.

**Comment by alfton** [comment]
it seems your are right, the bot have to stay in town for the exit delay to work.. So, we just have to find out how to make the bot go to town, or if there is a better solutions to this problem.

**Comment by AbbyRoad** [comment]
Didn't the command Town.doChores() cause all the bots to go back to town and then stay in town at the end of the run?

**Comment by alfton** [comment]
no, its not consistans, they do not leave at different time and at the time delay I have sett, sometime they enter town, but somtime they just leave the game at the same time as leader. Sometime they enter town, and just leave game at the same time as leader.. and that will get the bot in trouble.. fail to join game, and you get RD.

**Comment by AbbyRoad** [comment]
Do you have it set up so the leader and followers **all** go to town with the Town.doChores() command, and then use the delay exits on all of them (including the leader)?  Lowest delay would be on the leader, but it should be long enough to ensure all the bots get back to town before the leader exits?

**Comment by alfton** [comment]
yes, the followers are joining up in town, (I see that now), but, the problem stil exist, they leave to soon, the leecher unfortantly not join into town, and leaves at once baal i killed. I have set that follower (helper) should leave 20-30 sekund after leader, but sometime follower leave before leader, and visa versa. and the waiting time is no close to more then 1-2 sek. and less sometime, its not consistans.

**Comment by AbbyRoad** [comment]
In the Baal.js, BaalHelper.js and BaalAssistant.js files add the following lines beneath the Picket.pickItems(); line

Town.doChores();
delay(# in milliseconds);

For Baal, try 5000. (ie 5 sec)
For the BaalHelper, try 10000. (ie 10 sec)
For the BaalAssistant, try 15000 (ie 15 sec)

If these code work, then you can adjust the delay to whatever you want.  5 sec gaps should make it obvious if it working or not for all three.

Let me know if it is successful.

If you have more than one helper or more than one assistant, they will leave at the same time, but I don't think that will cause a FTJ.

**Comment by alfton** [comment]
I will try, and get back to you
unfortantly yes thos 2 got fail to join, becouse of leaving at same time, but the leader did not.

I guess blizzard have implentet some stuff here to detect unusually stuff.. So I think we need to make it harder to detect for blizzard that bot is at fot.. :)

**Comment by AbbyRoad** [comment]
This might be a better suggestion than the previous. Hopefully one or the other works.

1.  Increase the MinGameTime to a level that is higher than what a typical run should take for all the follower bots.  This will put all the followers into a wait mode after Baal is killed.
2.  Keep the Town.doChores() in the Baal file so the leader will go to town and hangout for a short time.  This ensures that all of the followers are in the wait mode.
3.  The QuitListDelay in each of the char files should then kick in once the leader leaves.

If the QuitListDelay function does require the followers to be in town in order to work - and if the followers don't automatically return to town in order to wait - then you'd need to keep the Town.doChores() in the BaalAssistant and BaalHelper files.  I'd probably then increase the QuitListDelay on the Leader to ensure that the followers had competed their chores and were in the wait mode before the leader left.

Please advise if any of these work.

**Comment by alfton** [comment]
its so strange behaviour of the followers, sometime they are going in to town with the leader, and pause awile, but sometime they just leave game at once baal is killed, and the leader stand in game goint to town, sell item, wait abit, and then leave,  ( i have 2 bot that follow leader, one that assist in killing, and one that just leache and "drinking koffee", i did try the delay thing afte the town.doChores, in baal, baalhelper, and baalassist. The thing with 2 bot leaving at same time, you can get away with it sometime, but if it happend to offend doring periode of time, I guess your follower get the fail to join game. unfortantly the minimun game lengt dalay, just send the bot out in lobby, and it wait abit longer before it make next game, no help.

**Comment by alfton** [comment]
Well, it work just nice enougth with one helper in game, the helper leave right after baal is killed, and leader goes to town, and sell item, and usaully wait abit before leave.. so untill we find the right cure for this, I lett my leecher have a wecations..

**Comment by AbbyRoad** [comment]
Well I've got good and bad news.  The good news, is that it is possible to get the delayed exit to work.  The bad news is that to do it, I changed too many things at once and I have yet to figure out what was the "critical" change.  So, tomorrow I'll need to do some file comparisons to figure it out.

In my current runs I have 3 bots running together and one just doing the chests quest.  At the end of baal, the three bots come back to town and sit there (that's what I need to figure out).  Once the fourth bot is finished it idles in town and once it hits the min game limit, it exits, and the other 3 bots leave based on the delayed quit levels in their files.

The team of three bots seem to sitting there until the max game limit is reached.  I found this out by adding a quest after baal, which only the leader did.  Once the leader left town, the idle time message appeared over the other two bots heads.  If I set the maximum run time limit on one of these two followers less than the minimum run time limit of the fourth bot, then it exits first, and the others then leave based on their delayed quit times.  So, this option would work if all 4 bots are doing the same thing,

To get the bots to go to town, initially I had been using the "Town.goToTown(5);" command in the Baal bot folder, but removing it did not change the bot behaviour (ie. they still went to town, versus leaving the game, even though the run had exceeded the mingame length for those three bots.

More tomorrow.  Hopefully there is enough here to allow someone to figure out the solution.  Sigh...

**Comment by alfton** [comment]
but you did not experience any fail to join when you got the bot to leave at different time delay?

**Comment by AbbyRoad** [comment]
Once I increased the delay to make the new game time to 15 sec and the delay to leave the game for the leader to 5 secs (made changes at the same time), I've seen no issues (ie. needed to address a fail to make, vs fail to join, issue).  As the games at 12 min long, it will take a several hours to confirm no fail to make/join issues. 

System seems to be really slow at making new games today.  Don't know if the issue is on their end or mine.

**Comment by AbbyRoad** [comment]
Over a 16 hour period, I have had 6 fail to create; 0 fail to join, 1 very short realm down (3 bots affected - leader and 2 followers) and 1 windows has unexpectedly exited event.  My games are 12 min start to end (12:20 - 12:40 min start to start), which puts me just under 20 runs/hr on a single IP (4 bots/run).

After a fail to create, I idle for 6 minutes before attempting to make the next game - which has always been successfully made.

The minimum game length is set by one of my follower bots, and it is set to be higher than the expected length of the run.  In my case, the runs typically take 10 - 10.5 min, so the bots idle in town for 1.5 - 2 min until minimum game length is reached, and then this follower leaves first, and the leader and remaining followers leave based on the staggers I have set.  (As a confirmation test - When I set one of the follower minimum game length times to less than what the run actually takes, then all the bots leave once baal is dead, with no delays.)

To be honest I think all I have done is switch a FTJ to a FTC, but this is a net positive as it prevents the runs/hr count from cycling up and thus it keeps me under 20 bot entries/hr.  (The leader is the second bot to leave.)

I hope this helps.

**Comment by alfton** [comment]
that is indead good news,, what did you have to do to make this work, changes i whats scripts?

**Comment by AbbyRoad** [comment]
**In D2BotFollow.dbj** - 

- Added JoinDelay for bots under line 47 (ie. beneath - // Put your lines under this one....)  (Note - I don't know if this helps, but it should not hurt)

I increased the following values : 
- JoinRetryDelay: 270,  (To prevent follower FTJs from running up the run count)
- RealmDownDelay: 6,

**In your leader and follower bot files change**:

- Config.MinGameTime = XXX, in **one follower** to the game length you wish (but the actual time needed to complete the run must be less than this **to ensure the bots are idling**) ; and in the **leader and other followers** to a time which is > the game length you wish. 
- Config.MaxGameTime to any value higher than the min game time used in the same bot file.
- In the line below the Config.QuitListMode command, add:
Config.QuitListDelay = [x,y];  x & y are the min and max delay (in seconds) you want for the delay with that bot's exit by.

**In D2BotLead.dbj file, I increased the following values** : 

- MinGameTime: 360,  (To prevent follower FTJs from running up the run count)
- CreateGameDelay: 15,
- FTJDelay: 360,
- RealmDownDelay: 6,
 
Good Luck!

**Comment by alfton** [comment]
ty for your effort AbbeyRoad, your solutions, did make me find another on to that did work flawlessly, and you dont need to add any code, or the exit delay code (that did not work as it should)
so my solutions is as following:

in the config file for your bot.

First off, turn off the leave game with leader, 
Config.Leader = "leadername"; // Leader's ingame character name. Leave blank to try auto-detection (works in AutoBaal, Wakka, MFHelper)
	Config.QuitList = [""]; // List of character names to quit with. Example: Config.QuitList = ["MySorc", "MyDin"];

(you dont need the quitlistdelay)

at line 432, you have the mingametime:
Config.MinGameTime = 1500; // Min game time in seconds. Bot will TP to town and stay in game if the run is completed before.
	Config.MaxGameTime = 1560; // Maximum game time in seconds. Quit game when limit is reached.

what I did was following,
my run is arround 24 min, so I put for the leader 25 min stall
and for the helper abit longer then the leader,
and at last the leecher (the one who does nothing, and just drink koffee wile the big boys do all the dirty work) even longer MinGameTime

Found this out to be rock sollid.
